### PR TITLE
Issue/8070 > master

### PR DESCRIPTION
**Pull request by davidgrindmill**
Configure the product uploading flow:

- Create a “Product uploader“ Drupal role, only this role should be able to create Product contents.
- Add the “Product uploader” role to the user upon registration when the user’s email address contains a “+productuploader” fragment (e.g. johndoe+productuploaded@gmail.com ).
- If the user changes email address on the profile page then the role should be either added or removed depending on the email address.

### CODE REVIEW

File: web/modules/custom/product_uploader/product_uploader.module [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+/**
+ * @file
+ * Contains hook for Product Uploader module.
+ */
+
+declare(strict_types=1);
```
Comment (lekob) [inline web/modules/custom/product_uploader/product_uploader.module:8]: This should be moved to right after the opening php tags.

File: web/modules/custom/product_uploader/product_uploader.info.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+name: Product Uploader
+type: module
+description: Assign or remove product uploader role
+package: Custom
+core: 8.x
```
Comment (lekob) [inline web/modules/custom/product_uploader/product_uploader.info.yml]: Use `core_version_requirement` key instead.

https://www.drupal.org/node/3070687

**Review by lekob — CHANGES_REQUESTED** [review]

File: web/modules/custom/product_uploader/product_uploader.module [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+use Drupal\Core\Entity\EntityInterface;
+
+/**
+ * Implements hook_entity_presave().
+ */
+function product_uploader_entity_presave(EntityInterface $entity) {
```
Comment (gkurucz) [inline web/modules/custom/product_uploader/product_uploader.module]: As this hook is called when any kind of entity is created or update, you should rather use

https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21entity.api.php/function/hook_ENTITY_TYPE_presave/8.8.x

This only gets called when the entity of an entity type is created or updated. This way you can also remove the outermost condition from the hook.

File: web/modules/custom/product_uploader/product_uploader.module [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  if ($entity->bundle() === 'user') {
+    $mail = $entity->get('mail')->value;
+    if (strpos($mail, '+productuploader')) {
+      $entity->addRole('product_uploader');
```
Comment (gkurucz) [inline web/modules/custom/product_uploader/product_uploader.module]: In the else block you check if this user has the `product_uploader` role, but you don't check it here. Is not necessary to check for the role, or did you miss something?

**Review by gkurucz — CHANGES_REQUESTED** [review]

File: web/modules/custom/product_uploader/product_uploader.info.yml (hunk 2 above)
Comment (davidgrindmill) [reply web/modules/custom/product_uploader/product_uploader.info.yml]: Fixed.

File: web/modules/custom/product_uploader/product_uploader.module (hunk 4 above)
Comment (davidgrindmill) [reply web/modules/custom/product_uploader/product_uploader.module]: It's not part of the task so I removed it.

File: web/modules/custom/product_uploader/product_uploader.module (hunk 3 above)
Comment (davidgrindmill) [reply web/modules/custom/product_uploader/product_uploader.module]: I'm going to need some help with this.

**Review by gkurucz — APPROVED** [review]

**Review by lekob — APPROVED** [review]

**Pull request merged**